llvm/include/llvm/ABI/Types.h

```diff
@@ -0,0 +1,538 @@
+  }
+
+  bool isVoid() const { return Kind == TypeKind::Void; }
+  bool isInteger() const { return Kind == TypeKind::Integer; }
+  bool isFloat() const { return Kind == TypeKind::Float; }
+  bool isPointer() const { return Kind == TypeKind::Pointer; }
+  bool isArray() const { return Kind == TypeKind::Array; }
+  bool isVector() const { return Kind == TypeKind::Vector; }
+  bool isStruct() const { return Kind == TypeKind::Struct; }
+  bool isMemberPointer() const { return Kind == TypeKind::MemberPointer; }
+  bool isComplex() const { return Kind == TypeKind::Complex; }
+};
+
+class VoidType : public Type {
+public:
+  VoidType() : Type(TypeKind::Void, TypeSize::getFixed(0), Align(1)) {}
+
+  static bool classof(const Type *T) { return T->getKind() == TypeKind::Void; }
+};
+
+class ComplexType : public Type {
+public:
+  ComplexType(const Type *ElementType, uint64_t SizeInBits, Align Alignment)
+      : Type(TypeKind::Complex, TypeSize::getFixed(SizeInBits), Alignment),
+        ElementType(ElementType) {}
+
+  const Type *getElementType() const { return ElementType; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Complex;
+  }
+
+private:
+  const Type *ElementType;
+};
+
+class IntegerType : public Type {
+private:
+  bool IsSigned;
+  bool IsBoolean;
+  bool IsBitInt;
+  uint64_t ABISizeInBits;
+
+public:
+  IntegerType(uint64_t BitWidth, uint64_t ABIWidth, Align Align, bool Signed,
+              bool IsBool = false, bool BitInt = false)
+      : Type(TypeKind::Integer, TypeSize::getFixed(BitWidth), Align),
+        IsSigned(Signed), IsBoolean(IsBool), IsBitInt(BitInt),
+        ABISizeInBits(ABIWidth) {}
+
+  bool isSigned() const { return IsSigned; }
+  bool isBool() const { return IsBoolean; }
+  bool isBitInt() const { return IsBitInt; }
+  uint64_t getLogicalBitWidth() const {
+    return getSizeInBits().getFixedValue();
+  }
+  uint64_t getABISizeInBits() const { return ABISizeInBits; }
+  bool isABIWidened() const { return getLogicalBitWidth() != ABISizeInBits; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Integer;
+  }
+};
+
+class FloatType : public Type {
+private:
+  const fltSemantics *Semantics;
+
+public:
+  FloatType(const fltSemantics &FloatSemantics, Align Align)
+      : Type(TypeKind::Float,
+             TypeSize::getFixed(APFloat::getSizeInBits(FloatSemantics)), Align),
+        Semantics(&FloatSemantics) {}
+
+  const fltSemantics *getSemantics() const { return Semantics; }
+  static bool classof(const Type *T) { return T->getKind() == TypeKind::Float; }
+};
+
+class PointerLikeType : public Type {
+protected:
+  unsigned AddrSpace;
+  PointerLikeType(TypeKind K, TypeSize Size, Align Align, unsigned AS = 0)
+      : Type(K, Size, Align), AddrSpace(AS) {}
+
+public:
+  virtual ~PointerLikeType() = default;
+  unsigned getAddrSpace() const { return AddrSpace; }
+  virtual bool isMemberPointer() const = 0;
```

Comment:
Don't use a virtual method for this. You can implement it by checking getKind().